plutarch-test/src/Plutarch/TryFromSpec.hs

```diff
@@ -0,0 +1,482 @@
+          l4 = toDatadList [6, 8, 8, 9, 10]
+      "concatenate two lists, legal"
+        @| validator # pforgetData l1 # pforgetData l2 # validContext0 @-> psucceeds
+      "concatenate two lists, illegal (list too short)"
+        @| validator # pforgetData l1 # pforgetData l3 # validContext0 @-> pfails
+      "concatenate two lists, illegal (wrong elements in list)"
+        @| validator # pforgetData l1 # pforgetData l4 # validContext0 @-> pfails
+      "concatenate two lists, illegal (more than one output)"
+        @| validator # pforgetData l1 # pforgetData l2 # invalidContext1 @-> pfails
+
+------------------- Checking deeply, shallowly and unwrapping ----------------------
+
+checkDeep ::
+  forall (target :: PType) (actual :: PType).
+  ( PTryFrom PData (PAsData target)
+  , PIsData actual
+  , PIsData target
+  ) =>
+  ClosedTerm (PAsData actual) ->
+  ClosedTerm (PAsData target)
+checkDeep t = unTermCont $ fst <$> TermCont (ptryFrom $ pforgetData t)
+
+checkDeepUnwrap ::
+  forall (target :: PType) (actual :: PType).
+  ( PTryFrom PData (PAsData target)
+  , PIsData actual
+  , PIsData target
+  ) =>
+  ClosedTerm (PAsData actual) ->
+  ClosedTerm (PTryFromExcess PData (PAsData target))
+checkDeepUnwrap t = unTermCont $ snd <$> TermCont (ptryFrom @PData @(PAsData target) $ pforgetData t)
+
+sampleStructure :: Term _ (PAsData (PBuiltinList (PAsData (PBuiltinList (PAsData (PBuiltinList (PAsData PInteger)))))))
+sampleStructure = pdata $ psingleton #$ pdata $ psingleton #$ toDatadList [1 .. 100]
+
+-- | PData serves as the base case for recursing into the structure
+partialCheck :: Term _ (PAsData (PBuiltinList (PAsData (PBuiltinList PData))))
+partialCheck =
+  let dat :: Term _ PData
+      dat = pforgetData sampleStructure
+   in unTermCont $ fst <$> TermCont (ptryFrom dat)
+
+fullCheck :: Term _ (PAsData (PBuiltinList (PAsData (PBuiltinList (PAsData (PBuiltinList (PAsData PInteger)))))))
+fullCheck = unTermCont $ fst <$> TermCont (ptryFrom $ pforgetData sampleStructure)
+
+------------------- Example: untrusted Redeemer ------------------------------------
+
+newtype PNatural (s :: S) = PMkNatural (Term s PInteger)
+  deriving (PlutusType, PIsData, PEq, POrd) via (DerivePNewtype PNatural PInteger)
+
+-- | partial
+pmkNatural :: Term s (PInteger :--> PNatural)
+pmkNatural = plam $ \i -> pif (i #< 0) (ptraceError "could not make natural") (pcon $ PMkNatural i)
+
+instance PTryFrom PData (PAsData PNatural) where
+  type PTryFromExcess PData (PAsData PNatural) = PNatural
+  ptryFrom opq = runTermCont $ do
+    (wrapped, unwrapped) <- TermCont $ ptryFrom @PData @(PAsData PInteger) opq
+    ver <- tcont $ plet $ pmkNatural # unwrapped
+    pure $ (punsafeCoerce wrapped, ver)
+
+validator :: Term s PValidator
+validator = phoistAcyclic $
+  plam $ \dat red ctx -> unTermCont $ do
+    (_, trustedRedeemer) <- TermCont $ ptryFrom @PData @(PAsData (PBuiltinList (PAsData PNatural))) red
```

Comment:
Can we use a custom data type that is PIsDataRepr for both redeemer and datum? That would make the example much more convincing

Reply:
I will try that.

Reply:
I have added an example that utilizes a custom datatype and shows how deriving `PTryFrom` can work with those.